tests/collaborative/collaborative_monkey_party.test.ts

```diff
@@ -0,0 +1,310 @@
+      action.user = newUsers[action.user.getters.getClient().name.toLowerCase()];
+    }
+  }
+  return run(network, [alice, bob, charlie], actions);
+}
+
+function minimizeFailingCommands(actions: UserAction[][]) {
+  // try removing concurrent actions chunk by chunk
+  let reduced = true;
+  while (reduced) {
+    reduced = false;
+    for (let chunkIndex = 0; chunkIndex < actions.length; chunkIndex++) {
+      const testCase = actions.slice(0, chunkIndex).concat(actions.slice(chunkIndex + 1));
+
+      const { fail } = rerunTest(testCase);
+      if (fail) {
+        actions = testCase;
+        reduced = true;
+      } else {
+        // reduce the chunk command-by-command
+        let chunkReduced = true;
+        while (chunkReduced) {
+          chunkReduced = false;
+          const chunk = actions[chunkIndex];
+          for (let cmdIndex = 0; cmdIndex < chunk.length; cmdIndex++) {
+            const testChunkCase = [...actions];
+            testChunkCase[chunkIndex] = chunk.slice(0, cmdIndex).concat(chunk.slice(cmdIndex + 1));
+            const { fail } = rerunTest(testChunkCase);
+            if (fail) {
+              actions = testChunkCase;
+              chunkReduced = true;
+            }
+          }
+        }
+      }
+    }
+  }
+  return actions;
+}
+
+function areSynced(users: Model[]): boolean {
+  for (let i = 0; i < users.length - 1; i++) {
+    if (!deepEquals(users[i].exportData(), users[i + 1].exportData())) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function evaluationsAreSynced(users: Model[]): boolean {
+  for (let i = 0; i < users.length - 1; i++) {
+    for (const sheetId of users[i].getters.getSheetIds()) {
+      const sheetZoneXc = zoneToXc(users[i].getters.getSheetZone(sheetId));
+      const valuesUserA = getRangeValues(users[i], sheetZoneXc, sheetId);
+      const valuesUserB = getRangeValues(users[i + 1], sheetZoneXc, sheetId);
+      if (!deepEquals(valuesUserA, valuesUserB)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function run(network: MockTransportService, users: Model[], actions: UserAction[][]) {
+  const executed: UserAction[][] = [];
+  for (const commandGroup of actions) {
+    const concurrentlyExecuted: UserAction[] = [];
+    executed.push(concurrentlyExecuted);
+    try {
+      network.concurrent(() => {
+        for (const { command, user } of commandGroup) {
+          const result = user.dispatch(command.type, command);
+          if (result.isSuccessful) {
+            concurrentlyExecuted.push({ command, user });
+          }
+        }
+      });
+      expect(evaluationsAreSynced(users)).toBe(true);
+    } catch (e) {
+      console.error(e);
+      concurrentlyExecuted.push(...commandGroup);
+      return {
+        fail: true,
+        executedActions: executed,
+      };
+    }
+  }
+  return {
+    fail: !areSynced(users),
+    executedActions: executed,
+  };
+}
+
+function randomizeCommandsPayload(commands: CoreCommand[]) {
```

Comment:
I would put the helpers to generate random commands in the `revision_party.ts`  file to clean up this one a bit